WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs

```diff
@@ -76,9 +76,49 @@ private void StepInputRegistrationPhase()
 				}
 				else
 				{
-					round.SetPhase(Phase.ConnectionConfirmation);
+					var areAllTxoUnspent = await CheckTxoSpendStatusAsync(round, cancel).ConfigureAwait(false);
+					if (areAllTxoUnspent)
+					{
+						round.SetPhase(Phase.ConnectionConfirmation);
+					}
+					else
+					{
+						round.SetPhase(Phase.Ended);
+						round.LogInfo($"The round was disrupted by users who spent their coins.");
+					}
+				}
+			}
+		}
+
+		private async Task<bool> CheckTxoSpendStatusAsync(Round round, CancellationToken cancellationToken)
+		{
+			IEnumerable<(Alice Alice, Task<GetTxOutResponse?> Status)> GetTxoSpendingStatus()
+			{
+				foreach (var alice in round.Alices)
+				{
+					var input = alice.Coin.Outpoint;
+					var getTxOutTask = Rpc.GetTxOutAsync(input.Hash, (int)input.N, includeMempool: true);
+					yield return (alice, getTxOutTask);
+				}
+			}
+
+			var allTxoAreUnspent = true;
+			var batchedRpc = Rpc.PrepareBatch();
+			var spendStatusCheckingTasks = GetTxoSpendingStatus().ToArray();
+
+			cancellationToken.ThrowIfCancellationRequested();
+			await batchedRpc.SendBatchAsync().ConfigureAwait(false);
+
+			foreach (var (alice, task) in spendStatusCheckingTasks)
+			{
+				var status = await task.ConfigureAwait(false);
+				if (status is null)
```

Comment:
FMI: I wonder here about the precise meaning here. It seems to me that based on `gettxout` [implementation](https://github.com/bitcoin/bitcoin/blob/548ca1d3a577cf53da93572140327256a2035918/src/rpc/blockchain.cpp#L1245), this condition is true if either the UTXO is spent *or* the UTXO does not exist at all. Do I understand correctly?

(https://github.com/bitcoin/bitcoin/issues/18476)

Reply:
Yes, this method returns true only if the TXO is found in the UTXO database and there is no transaction spending it in the mempool. In other words, it returns true only if thet TXO is really a UTXO.

Reply:
Hm, I don't understand it. :( I understand that `gettxout` returns `null` if the UTXO is either spent or it does not exist - NBitcoin actually documents that: https://github.com/MetacoSA/NBitcoin/blob/fb43cb426132e9ab4bf6594724ceb603e15418b9/NBitcoin/RPC/RPCClient.cs#L1693

Or is there some previous code that makes sure that the UTXO really exists?

Sorry, if it is downright obvious or I'm being dumb here. Just want to understand it.

Reply:
The root of the problem is that the names are wrong. Lets start with the concept of `utxo`, it means `Unpent Transaction Output`. So, the question/comments about whether a utxo can be spent or unspent makes no sense.

The other confusing name is `gettxout` that makes us think that it return info about a `txo` when in reality it return info about `utxo` only, a better name would be `getunspenttxout`.

Finally, bitcoin nodes maintain a database with all the `utxo` that's called `utxoset`. The rpc method `gettxout` check if the `txo` we pass is in the `utxoset` (if it is in the `utxoset` it means it is a `utxo`). However, it could happen that the `txo` is in the `utxoset` but there is a transaction waiting in the mempool that is spending that `utxo` so, in that case, is that a unspent coin or is it a spent coin that was not confirmed yet? Well, for our purpose we consider that that is a spent coin. In fact, it is an attack because it means that someone registered a coin in a coinjoin round and immediately spent it.